README.md
# sparkline

[![Build Status](https://travis-ci.org/Rowno/sparkline.svg?branch=master)](https://travis-ci.org/Rowno/sparkline)
[![Dependency Status](https://david-dm.org/Rowno/sparkline/status.svg)](https://david-dm.org/Rowno/sparkline)

Lightweight React sparklines ✨ 📈

![Example sparkline](example1.png)


## Install

```sh
yarn add @rowno/sparkline
# or
npm install --save @rowno/sparkline
```


## Example

```js
function Spark() {
  const lines = [
    {
      values: [789, 880, 676, 200, 890, 677, 900],
      colors: {
        area: 'rgba(217, 227, 237, 0.5)',
        line: '#193652'
      }
    }, {
      values: [354, 456, 200, 566, 344, 467, 545],
      colors: {
        area: 'rgba(199, 228, 255, 0.5)',
        line: '#004585'
      }
    }
  ]

  return (
    <Sparkline
      width={56}
      height={12}
      lines={lines}
      />
  )
}
```

Outputs: ![Example output](example2.png)


## Properties

```js
{
  width: 56,
  height: 12,
  lines: [{
    values: [789, 880, 676],
    colors: {
      area: 'rgba(217, 227, 237, 0.5)',
      line: '#193652'
    },
    title: 'Allowed events',
    key: 'allowed'
  }]
}
```

### width

Type: `number` (required)

Width of the sparkline.

### height

Type: `number` (required)

Height of the sparkline.

### lines

Type: `array<object>` (required)

Objects defining the lines to draw.

#### values

Type: `array<number>` (required)

Numbers that make up the data points of the line.

#### colors

Type: `object`

Custom colors for the line.

##### area

Type: `string`

Color of the line's filled in area.

##### line

Type: `string`

Color of the line's stroke.

#### title

Type: `string`

`title` of the line. Shown as a tooltip in the browser.

#### key

Type: `any`

Unique React `key` of the line.


## License

sparkline is released under the ISC license.

Copyright © 2017, Roland Warmerdam.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
This project follows [Semantic Versioning](http://semver.org).

## 1.0.0 / 2017-06-24
 - initial release.


lib/index.test.js
import test from 'ava'
import {add} from './index'

test('adds domain to url', t => {
  t.is(add(2, 3), 5)
})


src/index.test.js
import test from 'ava'
import React from 'react'
import render from 'react-test-renderer'
import {shallow} from 'enzyme'
import Sparkline from './index'

const values = [
  354,
  456,
  200,
  566,
  344,
  467,
  545
]
const width = 56
const height = 12

test('snapshot', t => {
  const component = (
    <Sparkline
      width={width}
      height={height}
      lines={[{values, title: 'Title 1'}, {values, title: 'Title 2'}]}
      />
  )
  t.snapshot(render.create(component).toJSON())
})

test('width', t => {
  const width = 123
  const component = (
    <Sparkline
      width={width}
      height={height}
      lines={[{values}]}
      />
  )
  const wrapper = shallow(component)
  t.is(wrapper.prop('width'), width)
  t.is(wrapper.prop('viewBox'), `0 0 ${width} ${height}`)
})

test('height', t => {
  const height = 123
  const component = (
    <Sparkline
      width={width}
      height={height}
      lines={[{values}]}
      />
  )
  const wrapper = shallow(component)
  t.is(wrapper.prop('height'), height)
  t.is(wrapper.prop('viewBox'), `0 0 ${width} ${height}`)
})

test('colors', t => {
  const colors = {area: '#e4f7ed', line: '#2B6B4C'}
  const component = (
    <Sparkline
      width={width}
      height={height}
      lines={[{values, colors}]}
      />
  )
  const paths = shallow(component).find('path')
  t.is(paths.first().prop('fill'), colors.area)
  t.is(paths.last().prop('stroke'), colors.line)
})

test('title', t => {
  const title = 'Line title'
  const component = (
    <Sparkline
      width={width}
      height={height}
      lines={[{values, title}]}
      />
  )
  const wrapper = shallow(component)
  t.is(wrapper.find('path title').text(), title)
})

test('multiple lines', t => {
  const component = (
    <Sparkline
      width={width}
      height={height}
      lines={[{values}, {values}]}
      />
  )
  const wrapper = shallow(component)
  t.is(wrapper.find('g').length, 2)
  t.is(wrapper.find('path').length, 4)
})


src/index.js
import React, {PureComponent} from 'react'
import PropTypes from 'prop-types'
import {area, line, curveMonotoneX} from 'd3-shape'
import {scaleLinear} from 'd3-scale'
import {max} from 'd3-array'

const curveFunction = curveMonotoneX
const defaultColors = {
  area: 'rgba(199, 228, 255, 0.5)',
  line: '#004585'
}

export default class Sparkline extends PureComponent {
  static displayName = 'Sparkline'
  static propTypes = {
    lines: PropTypes.arrayOf(
      PropTypes.shape({
        values: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired,
        colors: PropTypes.shape({
          area: PropTypes.string.isRequired,
          line: PropTypes.string.isRequired
        }),
        title: PropTypes.string,
        key: PropTypes.any
      })
    ).isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired
  }

  render() {
    const {lines, width, height} = this.props
    const maxX = max(lines, s => s.values.length - 1)
    const maxY = max(lines, s => max(s.values))
    const x = scaleLinear().domain([0, maxX]).range([0, width])
    // Set range to 1 to make room for the line stroke
    const y = scaleLinear().domain([0, maxY]).range([height, 1])

    const areaFunction =
      area()
      .x((d, i) => x(i))
      .y0(height)
      .y1(d => y(d))
      .curve(curveFunction)

    const lineFunction =
      line()
      .x((d, i) => x(i))
      .y(d => y(d))
      .curve(curveFunction)

    return (
      <svg width={width} height={height} viewBox={`0 0 ${width} ${height}`}>
        {lines.map((line, index) => {
          const colors = {
            ...defaultColors,
            ...line.colors
          }
          return (
            <g key={line.key || index}>
              <path d={areaFunction(line.values)} fill={colors.area}>
                {line.title && <title>{line.title}</title>}
              </path>
              <path d={lineFunction(line.values)} stroke={colors.line} fill="none"/>
            </g>
          )
        })}
      </svg>
    )
  }
}


src/index.story.js
import React from 'react'
import {storiesOf} from '@storybook/react'
import Sparkline from './index'

const width = 56
const height = 12
const values1 = [
  354,
  456,
  200,
  566,
  344,
  467,
  545
]
const values2 = [
  789,
  880,
  676,
  200,
  890,
  677,
  900
]

storiesOf('Sparkline', module)
  .add('basic', () => {
    const lines = [{
      values: values1,
      title: 'Allowed events'
    }]
    return (
      <Sparkline
        width={width}
        height={height}
        lines={lines}
        />
    )
  })
  .add('lots of data', () => {
    const lines = [{
      values: values1.concat(values1).concat(values1).concat(values1)
    }]
    return (
      <Sparkline
        width={width}
        height={height}
        lines={lines}
        />
    )
  })
  .add('custom colors', () => {
    const lines = [{
      values: values1,
      colors: {area: '#e4f7ed', line: '#2B6B4C'}
    }]
    return (
      <Sparkline
        width={width}
        height={height}
        lines={lines}
        />
    )
  })
  .add('square', () => {
    return (
      <Sparkline
        width={width}
        height={width}
        lines={[{values: values1}]}
        />
    )
  })
  .add('multiple sparklines', () => {
    const lines = [
      {
        values: values2,
        colors: {
          area: 'rgba(217, 227, 237, 0.5)',
          line: '#193652'
        },
        title: 'Denied events'
      }, {
        values: values1,
        colors: {
          area: 'rgba(199, 228, 255, 0.5)',
          line: '#004585'
        },
        title: 'Allowed events'
      }
    ]
    return (
      <Sparkline
        width={width}
        height={height}
        lines={lines}
        />
    )
  })
  .add('many sparks, so fast', () => {
    const sparklines = []
    for (let i = 0; i < 200; i++) {
      sparklines.push(
        <Sparkline
          key={i} // eslint-disable-line react/no-array-index-key
          width={width}
          height={height}
          lines={[{values: values1}]}
          />
      )
    }
    return (
      <div>
        {sparklines}
      </div>
    )
  })
